### PR TITLE
Authentication changes

### LINKED ISSUE #15: Collision with `pusher` gem
We'd like to use Chatkit on Rails project, however when we include `chatkit-server-ruby`, we hit the following error in `pusher-platform`:

```
~/ruby/gems/2.4.0/gems/pusher-platform-0.6.0/lib/pusher-platform/common.rb:2:
in `<module:Pusher>': superclass mismatch for class Error (TypeError)
```

I think this is because of a collision between `Pusher::Error` on both gems:

**pusher-platform**
https://github.com/pusher/pusher-platform-ruby/blob/b077a70b1f911e90641bc6ec5406a5febb76f55f/lib/pusher-platform/common.rb

**pusher**
https://github.com/pusher/pusher-http-ruby/blob/adf04980682264050e70b99dd39dc8b82af2390a/lib/pusher.rb#L18

Is this correct, and do you know of a workaround?

### PR DESCRIPTION
**Pull request by hamchapman**
### What?

- Rename `authenticate` to `authenticate_with_request`
- `authenticate` (now `authenticate_with_request`) no longer returns a status code. It will either return a `Hash` containing the required authentication payload or will raise an error
- Changed module name from `Pusher` to `PusherPlatform` (fixes #15)
- Fixes some `require`s 
- Add some examples 

### Why?

- Makes it more consistent with Node SDK 

----

- [x] README updated if you changed the API?

----

CC @pusher/sigsdk

### CODE REVIEW

File: auth_server_example/sample.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+require 'sinatra'
+require 'json'
+require 'cgi'
+require_relative '../lib/pusher-platform'
+
+# Get these from the Dashbaord
+instance = PuPusherPlatformsher::Instance.new(
```
Comment (pl) [inline auth_server_example/sample.rb]: Typo? Would be nice if those examples were run in CI.

File: lib/pusher-platform/error_with_status.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
+require_relative './common'
+
+module PusherPlatform
+  class ErrorWithStatus < Error
```
Comment (pl) [inline lib/pusher-platform/error_with_status.rb]: Why do we need another error type?

File: auth_server_example/sample.rb (hunk 1 above)
Comment (hamchapman) [reply auth_server_example/sample.rb]: Most certainly a typo - thanks.

We'll get to having CI someday, but not worth it right now IMO

File: lib/pusher-platform/error_with_status.rb (hunk 2 above)
Comment (hamchapman) [reply lib/pusher-platform/error_with_status.rb]: The `ErrorResponse` error fit with the case where you make a request and get back error info in the response (description, error type, etc).

This type is about returning an error with a suggested status code. It's only used by auth-related things so perhaps the error could be named differently: `AuthenticationError(WithStatus)`?

**Comment by hamchapman** [comment]
@pl am I good to merge this?

**Comment by hamchapman** [comment]
Updated to use `ErrorResponse` in auth errors.

Basically matches Node SDK (after this PR is merged https://github.com/pusher/pusher-platform-node/pull/50)

**Comment by pl** [comment]
Looks good to me now.

**Pull request merged**